### DISCUSSION TITLE
How to eliminate the error returned by the otUdpNewMessage function

### DISCUSSION THREAD

**Discussion by tuopu** [openthread/openthread#8360] [Q&A] [answered]
Hello, experts：
      when  a new thread has been created in Linux. The otUdpNewMessage function is used to create UDP data, and the otUdpSen function is used to send UDP frames. When the amount of data sent is large, the otUdpNewMessage will return the error code OT_ ERROR_ NO_ BUFS will always return this error later, and cannot continue to send data. Please ask how to recover when this error occurs again, so that the thread can continue to run and send data.

**Comment by jwhui** [comment, marked as answer]
It takes time for OpenThread to process message buffers. The IEEE 802.15.4 radio has a PHY data rate of 250 kbps. The best way is to simply allow OpenThread to process events and tasks for some time before attempting to call `otUdpNewMessage()` again.

**Comment by tuopu** [comment]
This test is conducted under the Linux simulation without using other hardware. The phenomenon is that at first, data can be sent to another simulation instance normally using UDP, but after a period of time, otUdpNewMessage cannot apply for space and cannot be recovered. The routine references the cli_ Udp.CPP, but the interface has changed from terminal input to serial input.

**Reply by jwhui** [reply]
Can you provide more detail so that someone can reproduce the issue?

**Reply by tuopu** [reply]
In Linux, create a new process, connect an external serial port, use it to send and receive data, use the otUdpNewMessage function to apply for space, use the otUdpSend to send data, and the openthread protocol stack stops the terminal window listening function (rval=poll (pollfd, sizeof (pollfd)/sizeof (* pollfd), 0);, In the openthread examples platforms simulation directory, create another instance to receive data, and use the serial port to send the program to send data to the instance continuously. The time interval I use is 50ms. In less than a minute, I can no longer send data. At this time, the otUdpNewMessage function prompts that the memory space cannot be applied, and writes the reference cli_ Udp.CPP.

**Reply by jwhui** [reply]
It would be helpful if you can provide more detail and visibility. For example:
- debug-level logs
- CLI commands
- any code written or modified
- any test scripts that reliably reproduce the issue

**Reply by tuopu** [reply]
//Protocol stack UDP processing
void  udp_init(void)
{
    otUdpOpen(Instance, &mSocket, HandleUdpReceive, this); 
}
void modbus_udp::udp_send(udp_buf  buf)
{
    otError                 error = OT_ERROR_NONE;
    message               = otUdpNewMessage(Instance,null);
    VerifyOrExit(message != nullptr, error = OT_ERROR_NO_BUFS);
    otMessageAppend(message, &buf,  drift_len);
    SuccessOrExit(error = otUdpSend(Instance, &mSocket, message, &messageInfo));
    message = nullptr;
exit:
    if (message != nullptr)
    {
        otMessageFree(message);
    }
}
//Create a new thread to receive serial port data and send it as UDP
void heckdata(void)
{
    while (1)
    {
        memset(&buf, 0, sizeof(buf));
        buf.len = serial_receive(_fd, buf.buf, sizeof(buf.buf), 500);
       udp_send(buf);
    }
}

**Reply by tuopu** [reply]
When the speed of calling the otUdpNewMessage function to create memory space exceeds 100ms each time, the program will crash. The error is that the memory is out of bounds.
The common mistakes are as follows：

Abnormal occurrence1：
void PriorityQueue::Dequeue(Message &aMessage)
{
    Message::Priority priority;
    Message *         tail;

    OT_ASSERT(aMessage.GetPriorityQueue() == this);

Abnormal occurrence 2：
        aChunk.SetBuffer(aChunk.GetBuffer()->GetNextBuffer());

        OT_ASSERT(aChunk.GetBuffer() != nullptr);
Abnormal occurrence 3：
 Message *      GetTail(void) { return static_cast<Message *>(mData); }

Abnormal occurrence4:
void MessagePool::Free(Message *aMessage)
{
    OT_ASSERT(aMessage->Next() == nullptr && aMessage->Prev() == nullptr);

**Reply by jwhui** [reply]
Are you calling OpenThread APIs from different threads of execution? Note that the OpenThread APIs do not provide any synchronization.

**Reply by tuopu** [reply]
Yes, I call sending functions on different threads, that is, when I call these functions, I must be in the same thread to run normally, and I cannot operate openthread functions across threads.
